### PR TITLE
User provided log quantities in gsd 

### PR DESCRIPTION
**Pull request by joaander**
## Description

Implement user-defined log quantities defined in glotzerlab/gsd#38. Users can now provide callbacks that provide custom log quantities in gsd frames. These quantities can be scalars, strings, per-particle, per-bond, etc... In a future release, OVITO will disover these data and make them available in the data pipeline.

Users provide data via callables that return numpy arrays:
```python
gsd = hoomd.dump.gsd(filename="trajectory.gsd", period=1000, group=hoomd.group.all(), overwrite=True)
gsd.log['particles/potential_energy'] = lambda step: numpy.array([p.net_energy for p in system.particles], dtype=numpy.float32)
gsd.log['particles/net_force'] = lambda step: numpy.array([p.net_force for p in system.particles], dtype=numpy.float32)
gsd.log['potential_energy'] = lambda step: numpy.array([log.query('potential_energy')], dtype=numpy.float64)
message = 'hello OVITO'
gsd.log['string'] = lambda step: numpy.array([message], dtype=numpy.dtype((bytes, len(message)+1))).view(dtype=numpy.int8)
gsd.log['bonds/tag'] = lambda step: numpy.array([b.tag for b in system.bonds], dtype=numpy.uint32)
hoomd.run(1)
```

`log` is a dictionary. You can remove some entries to prevent them from being written on later frames:
```
del gsd.log['bonds/tag']
del gsd.log['string']

hoomd.run(10e3)
```

## Motivation and Context

Users would like to store arbitrary information in gsd files and use that information in analysis and visualization pipelines.

Resolves: #423 

## How Has This Been Tested?

Added unit tests.

## Change log

```
* User-defined log quantities in ``dump.gsd`` 
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).
- [ ] (Glotzer group members): I am a member of the [hoomd-contributors team](https://github.com/orgs/glotzerlab/teams/hoomd-contributors/members).

### CODE REVIEW

**Comment by joaander** [comment]
OVITO 3.0.0-dev498 and newer read custom logged data from gsd files. Please review the PR so we can get this capability into HOOMD in the next release.

File: hoomd/dump.py [hunk 1]
```diff
@@ -505,21 +505,20 @@ class gsd(hoomd.analyze._analyzer):
...
-    the file. Additionally, only **dynamic** quantities are written to all frames, non-dynamic
-    quantities are only written to frame 0. The GSD schema defines that data not present in frame *i* is to
-    be read from frame 0. This makes every single frame of a GSD file fully specified and simulations
-    initialized with :py:func:`hoomd.init.read_gsd()` can select any frame of the file.
-
-    You can control what quantities are dynamic by category. ``property`` is always dynamic.
-    The categories listed in the **dynamic** will also be written out to every frame.
+    Write a simulation snapshot to the specified GSD file at regular intervals. GSD is capable of storing all particle
+    and bond data fields in hoomd, in every frame of the trajectory. This allows GSD to store simulations where the
+    number of particles, number of particle types, particle types, diameter, mass, charge, or anything is changing over
```
Comment (willzygmunt) [inline hoomd/dump.py]: anything **that** is changing over time

Comment (willzygmunt) [inline hoomd/dump.py]: anything **that** is changing over time

**Comment by joaander** [comment]
Thanks!

**Pull request merged**